what4/src/What4/Expr/Builder.hs

```diff
@@ -1101,8 +1148,95 @@ sbConcreteLookup sym arr0 mcidx idx
+-- from the lookup index.
+sliceArrayLookupUpdate ::
+  ExprBuilder t st fs ->
+  Expr t (BaseArrayType (d::>tp) range) ->
+  Ctx.Assignment (Expr t) (d::>tp) ->
+  IO (Expr t (BaseArrayType (d::>tp) range), Ctx.Assignment (Expr t) (d::>tp))
+sliceArrayLookupUpdate sym arr0 lookup_idx
+  | Just (ArrayMap _ _ entry_map arr) <- asApp arr0 =
+    case asConcreteIndices lookup_idx of
+      Just lookup_concrete_idx ->
+        case AUM.lookup lookup_concrete_idx entry_map of
+          Just val -> do
+            arr_base <- arrayUpdateBase sym arr
+            sliced_arr <- arrayUpdate sym arr_base lookup_idx val
+            return (sliced_arr, lookup_idx)
+          Nothing -> sliceArrayLookupUpdate sym arr lookup_idx
+      Nothing ->
+        return (arr0, lookup_idx)
+
+  | Just (CopyArray _w _a_repr dest_arr dest_begin_idx src_arr src_begin_idx len dest_end_idx _src_end_idx) <- asApp arr0 = do
+    p0 <- bvUle sym dest_begin_idx (Ctx.last lookup_idx)
+    p1 <- bvUlt sym (Ctx.last lookup_idx) dest_end_idx
+    case (asConstantPred p0, asConstantPred p1) of
+      (Just True, Just True) -> do
+        new_lookup_idx <- bvAdd sym src_begin_idx =<<
+          bvSub sym (Ctx.last lookup_idx) dest_begin_idx
+        sliceArrayLookupUpdate sym src_arr $ singleton new_lookup_idx
+      (Just False, _) ->
+        sliceArrayLookupUpdate sym dest_arr lookup_idx
+      (_, Just False) ->
+        sliceArrayLookupUpdate sym dest_arr lookup_idx
+      _ -> do
+        (sliced_dest_arr, sliced_dest_idx) <- sliceArrayLookupUpdate sym dest_arr lookup_idx
+        sliced_dest_begin_idx <- bvAdd sym dest_begin_idx =<<
+          bvSub sym (Ctx.last sliced_dest_idx) (Ctx.last lookup_idx)
+        sliced_arr <- arrayCopy sym sliced_dest_arr sliced_dest_begin_idx src_arr src_begin_idx len
+        return (sliced_arr, sliced_dest_idx)
+
+    -- A lookup in an array set returns the value when inside the set range
+  | Just (SetArray _w _a_repr arr begin_idx val len end_idx) <- asApp arr0 = do
+    p0 <- bvUle sym begin_idx (Ctx.last lookup_idx)
+    p1 <- bvUlt sym (Ctx.last lookup_idx) end_idx
+    case (asConstantPred p0, asConstantPred p1) of
+      (Just True, Just True) -> do
+        arr_base <- arrayUpdateBase sym arr
+        sliced_arr <- arrayUpdate sym arr_base lookup_idx val
+        return (sliced_arr, lookup_idx)
+      (Just False, _) ->
+        sliceArrayLookupUpdate sym arr lookup_idx
+      (_, Just False) ->
+        sliceArrayLookupUpdate sym arr lookup_idx
+      _ -> do
+        (sliced_arr, sliced_idx) <- sliceArrayLookupUpdate sym arr lookup_idx
+        sliced_begin_idx <- bvAdd sym begin_idx =<<
+          bvSub sym (Ctx.last sliced_idx) (Ctx.last lookup_idx)
+        sliced_arr <- arraySet sym sliced_arr sliced_begin_idx val len
+        return (sliced_arr, sliced_idx)
+
+    -- Lookups on mux arrays just distribute over mux.
+  | Just (BaseIte _ _ p x y) <- asApp arr0 = do
```

Comment:
Another case where exponential slowdown is possible.

Reply:
this is similar to the removed code
```
    -- Lookups on mux arrays just distribute over mux.
  | Just (BaseIte _ _ p x y) <- asApp arr0 = do
      xv <- sbConcreteLookup sym x mcidx idx
      yv <- sbConcreteLookup sym y mcidx idx
      baseTypeIte sym p xv yv
```
with the difference that the mux over arrays results in a lookup with mux over indices and a mux over arrays. I don't think this introduces a new exponential slowdown.

Reply:
It's true this behavior was basically already here.  Let's open a ticket to track this, and the other instance I mentioned so we don't forget this is a potential issue.